[One Workflow] Fix virustotal scanFileHash 404

talboren opened the pull request:
https://github.com/user-attachments/assets/c4f35e52-2ab5-498c-b6b2-187d21142ce1


## Problem

VirusTotal actions throw errors when the API returns 404 or other HTTP errors, causing workflows to fail. This is problematic because 404 responses are expected for unknown resources (e.g., hashes not in the database, unknown IPs). Users currently need workarounds like `on-failure: continue: true` to prevent workflow failures.

## Changes

**VirusTotal Connector (`virustotal.ts`)**
- Added optional `failOnError` parameter (defaults to `false`) to all actions: `scanFileHash`, `scanUrl`, `submitFile`, and `getIpReport`
- When `false`: catches HTTP errors and returns structured response with error details
- When `true`: throws descriptive error with HTTP status code (e.g., "Hash not found in VirusTotal database (HTTP 404)", "VirusTotal API request failed (HTTP 429)")
- Each action has context-specific error messages for better debugging

**Tests (`virustotal.test.ts`)**
- Added 12 test cases covering error scenarios across all actions
- Tests cover both `failOnError=false` and `failOnError=true` behaviors
- Tests validate descriptive error messages include HTTP status codes

## Example Usage

```yaml
# Default behavior - no workflow failure on 404
- name: virusTotalTest
  type: virustotal.scanFileHash
  connector-id: Virustotal
  with:
    hash: "c34c444b4c345faea68984b75b7743271646c8008afbd036a9d3ec3838ee6fd2"

# Throw descriptive error on failures
- name: virusTotalTest
  type: virustotal.scanFileHash
  connector-id: Virustotal
  with:
    hash: "c34c444b4c345faea68984b75b7743271646c8008afbd036a9d3ec3838ee6fd2"
    failOnError: true
```

**Response for hash not found (failOnError: false):**
```json
{
  "id": null,
  "attributes": null,
  "stats": null,
  "error": {
    "status": 404,
    "message": "Hash not found in VirusTotal database",
    "details": { ... }
  }
}
```

**Error thrown when failOnError: true:**
```
Error: Hash not found in VirusTotal database (HTTP 404)
```

Copilot committed:
Initial plan

Copilot committed:
Add failOnError parameter to VirusTotal scanFileHash action

- Add optional failOnError parameter (defaults to false)
- When failOnError is false, return error details instead of throwing
- When failOnError is true, preserve original behavior
- Add meaningful error messages for 404 responses
- Add comprehensive tests for error handling scenarios

Co-authored-by: talboren <68807791+talboren@users.noreply.github.com>

Copilot committed:
Apply failOnError to all VirusTotal actions and improve error messages

- Add failOnError parameter to scanUrl, submitFile, and getIpReport actions
- When failOnError is true, throw descriptive errors with HTTP status codes
- Error messages now include action-specific context (e.g., "Hash not found in VirusTotal database (HTTP 404)")
- Add comprehensive tests for all actions covering error scenarios
- Updated existing tests to validate new error message format

Co-authored-by: talboren <68807791+talboren@users.noreply.github.com>

talboren committed:
Merge branch 'main' of https://github.com/elastic/kibana into fix-virustotal-scanfilehash-404

talboren committed:
Refactor VirusTotal error handling and improve test coverage

- Introduced a centralized error handling function for VirusTotal API responses to streamline error management across actions.
- Updated existing actions to utilize the new error handling function, enhancing clarity and maintainability.
- Improved error messages to provide more context, including HTTP status codes and specific action-related details.
- Expanded test cases to cover new error handling scenarios and validate the updated error message formats.

shahargl reviewed (approved):
lgtm

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/376959)
* Commit: 1b38ce1461ec22c25daf8507b5ad42c90d6c72dc

### Failed CI Steps
* [Jest Tests #16](https://buildkite.com/elastic/kibana-pull-request/builds/376959#019b6a4d-1c45-4f78-8e24-09995fc1c6cd)
* [FTR Configs #25](https://buildkite.com/elastic/kibana-pull-request/builds/376959#019b6a4d-1c6f-4c34-8a16-db4ab5e97eff)
* [FTR Configs #25](https://buildkite.com/elastic/kibana-pull-request/builds/376959#019b6a85-e4c0-4109-bf70-dece160b1701)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/376959#019b6a4d-1c6f-4c34-8a16-db4ab5e97eff) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/376959/jobs/019b6a4d-1c6f-4c34-8a16-db4ab5e97eff/artifacts/019b6a85-ba7f-438e-be88-514767bf5357) FTR Configs #<span></span>25 / Actions execute single file connector actions error handling should successfully validate correct params for action
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/376959#019b6a85-e4c0-4109-bf70-dece160b1701) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/376959/jobs/019b6a85-e4c0-4109-bf70-dece160b1701/artifacts/019b6a8d-9fad-4764-99e7-d9f87011821f) FTR Configs #<span></span>25 / Actions execute single file connector actions error handling should successfully validate correct params for action
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/376959#019b6a4d-1c45-4f78-8e24-09995fc1c6cd) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/376959/jobs/019b6a4d-1c45-4f78-8e24-09995fc1c6cd/artifacts/019b6a85-7ee4-402b-ba8a-0e8dd5d2cebb) Jest Tests #<span></span>16 / VirusTotalConnector scanFileHash action should return error details for 404 response when failOnError is false (default)
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/376959#019b6a4d-1c45-4f78-8e24-09995fc1c6cd) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/376959/jobs/019b6a4d-1c45-4f78-8e24-09995fc1c6cd/artifacts/019b6a85-7ee2-44b0-aea4-afd9d969b746) Jest Tests #<span></span>16 / VirusTotalConnector scanFileHash action should return error details for other HTTP errors when failOnError is false
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/376959#019b6a4d-1c45-4f78-8e24-09995fc1c6cd) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/376959/jobs/019b6a4d-1c45-4f78-8e24-09995fc1c6cd/artifacts/019b6a85-7ee2-4069-bf6b-8c91498d19d8) Jest Tests #<span></span>16 / VirusTotalConnector submitFile action should return error details when submission fails and failOnError is false

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/376959","number":376959,"commit":"1b38ce1461ec22c25daf8507b5ad42c90d6c72dc"}],"number":376959}
buildkite-pr-comment-->

talboren committed:
Refactor VirusTotal error response structure and update tests

- Removed unused properties from the ErrorResult interface to streamline error handling.
- Updated test cases to reflect changes in the expected error response structure.
- Enhanced error messages for clarity in test assertions, ensuring consistency with the new error handling approach.

talboren committed:
Merge branch 'main' into fix-virustotal-scanfilehash-404

talboren committed:
Merge branch 'main' into fix-virustotal-scanfilehash-404
